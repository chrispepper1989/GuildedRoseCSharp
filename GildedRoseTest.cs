using System;
using System.IO;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using csharp;

namespace GildedRoseCSharp
{
  
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class GildedRoseTest
    {

        //generate a bunch of test cases
        public IList<Item> GenerateTestItems(ValidItems[] names, int[] sellIns, int[] qualities)
        {
            IList<Item> Items = new List<Item>
            {
            };
            foreach (var name in names)
            {
                foreach (var sellIn in sellIns)
                {
                    foreach (var quality in qualities)
                    {
                        var item = ItemHelper.CreateItem(name, sellIn, quality);
                        Items.Add( item);
                    }

                }

            }

            return Items;

        }

        //[TestCase(10)] //todo why doesnt this work with Approvals?
        public void TestUpdateQuality(int days)
        {

            //arrange
            ValidItems[] namesProper = new[]
            {
                ValidItems.AgedBrie, ValidItems.DexterityVest, ValidItems.ElixirOfMongoose,
                ValidItems.SulfurasHandOfRagnaros, ValidItems.ConjuredManaCake
            };
            var sellIns = new[] { 1, 5, 10, 2, 4 };
            var qualities = new[] { 0, 1, 2, 5, 29 };
            var Items = GenerateTestItems(namesProper, sellIns, qualities);
            GildedRose app = new GildedRose(Items);
            //act
            for (var i = 0; i < days; ++i)
            {
                app.UpdateQuality();
            }
            //assert
            Approvals.VerifyAll(Items, "All Items");

        }
        [Test]
        public void ConjuredManaCakeDegradesTwiceAsFastAsNormalItems()
        {
            //arrange
            var conitem = ItemHelper.CreateItem(ValidItems.ConjuredManaCake, 10, 100);
            var normalItem = ItemHelper.CreateItem(ValidItems.ElixirOfMongoose, 10, 100);

            IList<Item> Items = new List<Item>
            {
         
            };

            GildedRose app = new GildedRose(Items);

            IList<Item> ItemsOverTime = new List<Item>();
            //act
            for (var i = 0; i < 100; ++i)
            {
                conitem.SellIn -= 1;
                normalItem.SellIn -= 1;
                //normal item done twice
                normalItem.Quality =  GildedRose.GetNewStandardItemQuality(normalItem);
                normalItem.Quality = GildedRose.GetNewStandardItemQuality(normalItem);
                conitem.Quality = GildedRose.GetNewConjuredManaCakeQuality(conitem);
               
                //assert
                Assert.AreEqual(normalItem.Quality, conitem.Quality);
      
                ItemsOverTime.Add(new Item()
                {
                    Name = normalItem.Name,
                    Quality = normalItem.Quality,
                    SellIn = normalItem.SellIn

                });
                ItemsOverTime.Add(new Item()
                {
                    Name = conitem.Name,
                    Quality = conitem.Quality,
                    SellIn = conitem.SellIn

                });

            }
            Approvals.VerifyAll(ItemsOverTime, "Conjured Items ");

        }

        [Test]
         public void TestUpdateQuality_1Days()
         {
             TestUpdateQuality(1);
         }
         [Test]
         public void TestUpdateQuality_5Days()
         {
             TestUpdateQuality(5);
         }
         [Test]
         public void TestUpdateQuality_10Days()
         {
             TestUpdateQuality(10);
         }
         [Test]
         public void TestUpdateQuality_30Days()
         {
             TestUpdateQuality(30);
         }

    }
}
