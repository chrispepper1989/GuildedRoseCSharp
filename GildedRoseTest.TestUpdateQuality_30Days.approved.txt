All Items[0] = Aged Brie, -29, 50
All Items[1] = Aged Brie, -29, 50
All Items[2] = Aged Brie, -29, 50
All Items[3] = Aged Brie, -29, 50
All Items[4] = Aged Brie, -29, 50
All Items[5] = Aged Brie, -25, 50
All Items[6] = Aged Brie, -25, 50
All Items[7] = Aged Brie, -25, 50
All Items[8] = Aged Brie, -25, 50
All Items[9] = Aged Brie, -25, 50
All Items[10] = Aged Brie, -20, 50
All Items[11] = Aged Brie, -20, 50
All Items[12] = Aged Brie, -20, 50
All Items[13] = Aged Brie, -20, 50
All Items[14] = Aged Brie, -20, 50
All Items[15] = Aged Brie, -28, 50
All Items[16] = Aged Brie, -28, 50
All Items[17] = Aged Brie, -28, 50
All Items[18] = Aged Brie, -28, 50
All Items[19] = Aged Brie, -28, 50
All Items[20] = Aged Brie, -26, 50
All Items[21] = Aged Brie, -26, 50
All Items[22] = Aged Brie, -26, 50
All Items[23] = Aged Brie, -26, 50
All Items[24] = Aged Brie, -26, 50
All Items[25] = +5 Dexterity Vest, -29, 0
All Items[26] = +5 Dexterity Vest, -29, 0
All Items[27] = +5 Dexterity Vest, -29, 0
All Items[28] = +5 Dexterity Vest, -29, 0
All Items[29] = +5 Dexterity Vest, -29, 0
All Items[30] = +5 Dexterity Vest, -25, 0
All Items[31] = +5 Dexterity Vest, -25, 0
All Items[32] = +5 Dexterity Vest, -25, 0
All Items[33] = +5 Dexterity Vest, -25, 0
All Items[34] = +5 Dexterity Vest, -25, 0
All Items[35] = +5 Dexterity Vest, -20, 0
All Items[36] = +5 Dexterity Vest, -20, 0
All Items[37] = +5 Dexterity Vest, -20, 0
All Items[38] = +5 Dexterity Vest, -20, 0
All Items[39] = +5 Dexterity Vest, -20, 0
All Items[40] = +5 Dexterity Vest, -28, 0
All Items[41] = +5 Dexterity Vest, -28, 0
All Items[42] = +5 Dexterity Vest, -28, 0
All Items[43] = +5 Dexterity Vest, -28, 0
All Items[44] = +5 Dexterity Vest, -28, 0
All Items[45] = +5 Dexterity Vest, -26, 0
All Items[46] = +5 Dexterity Vest, -26, 0
All Items[47] = +5 Dexterity Vest, -26, 0
All Items[48] = +5 Dexterity Vest, -26, 0
All Items[49] = +5 Dexterity Vest, -26, 0
All Items[50] = Elixir of the Mongoose, -29, 0
All Items[51] = Elixir of the Mongoose, -29, 0
All Items[52] = Elixir of the Mongoose, -29, 0
All Items[53] = Elixir of the Mongoose, -29, 0
All Items[54] = Elixir of the Mongoose, -29, 0
All Items[55] = Elixir of the Mongoose, -25, 0
All Items[56] = Elixir of the Mongoose, -25, 0
All Items[57] = Elixir of the Mongoose, -25, 0
All Items[58] = Elixir of the Mongoose, -25, 0
All Items[59] = Elixir of the Mongoose, -25, 0
All Items[60] = Elixir of the Mongoose, -20, 0
All Items[61] = Elixir of the Mongoose, -20, 0
All Items[62] = Elixir of the Mongoose, -20, 0
All Items[63] = Elixir of the Mongoose, -20, 0
All Items[64] = Elixir of the Mongoose, -20, 0
All Items[65] = Elixir of the Mongoose, -28, 0
All Items[66] = Elixir of the Mongoose, -28, 0
All Items[67] = Elixir of the Mongoose, -28, 0
All Items[68] = Elixir of the Mongoose, -28, 0
All Items[69] = Elixir of the Mongoose, -28, 0
All Items[70] = Elixir of the Mongoose, -26, 0
All Items[71] = Elixir of the Mongoose, -26, 0
All Items[72] = Elixir of the Mongoose, -26, 0
All Items[73] = Elixir of the Mongoose, -26, 0
All Items[74] = Elixir of the Mongoose, -26, 0
All Items[75] = Sulfuras, Hand of Ragnaros, 1, 0
All Items[76] = Sulfuras, Hand of Ragnaros, 1, 1
All Items[77] = Sulfuras, Hand of Ragnaros, 1, 2
All Items[78] = Sulfuras, Hand of Ragnaros, 1, 5
All Items[79] = Sulfuras, Hand of Ragnaros, 1, 29
All Items[80] = Sulfuras, Hand of Ragnaros, 5, 0
All Items[81] = Sulfuras, Hand of Ragnaros, 5, 1
All Items[82] = Sulfuras, Hand of Ragnaros, 5, 2
All Items[83] = Sulfuras, Hand of Ragnaros, 5, 5
All Items[84] = Sulfuras, Hand of Ragnaros, 5, 29
All Items[85] = Sulfuras, Hand of Ragnaros, 10, 0
All Items[86] = Sulfuras, Hand of Ragnaros, 10, 1
All Items[87] = Sulfuras, Hand of Ragnaros, 10, 2
All Items[88] = Sulfuras, Hand of Ragnaros, 10, 5
All Items[89] = Sulfuras, Hand of Ragnaros, 10, 29
All Items[90] = Sulfuras, Hand of Ragnaros, 2, 0
All Items[91] = Sulfuras, Hand of Ragnaros, 2, 1
All Items[92] = Sulfuras, Hand of Ragnaros, 2, 2
All Items[93] = Sulfuras, Hand of Ragnaros, 2, 5
All Items[94] = Sulfuras, Hand of Ragnaros, 2, 29
All Items[95] = Sulfuras, Hand of Ragnaros, 4, 0
All Items[96] = Sulfuras, Hand of Ragnaros, 4, 1
All Items[97] = Sulfuras, Hand of Ragnaros, 4, 2
All Items[98] = Sulfuras, Hand of Ragnaros, 4, 5
All Items[99] = Sulfuras, Hand of Ragnaros, 4, 29
All Items[100] = Sulfuras, Hand of Ragnaros, 1, 0
All Items[101] = Sulfuras, Hand of Ragnaros, 1, 1
All Items[102] = Sulfuras, Hand of Ragnaros, 1, 2
All Items[103] = Sulfuras, Hand of Ragnaros, 1, 5
All Items[104] = Sulfuras, Hand of Ragnaros, 1, 29
All Items[105] = Sulfuras, Hand of Ragnaros, 5, 0
All Items[106] = Sulfuras, Hand of Ragnaros, 5, 1
All Items[107] = Sulfuras, Hand of Ragnaros, 5, 2
All Items[108] = Sulfuras, Hand of Ragnaros, 5, 5
All Items[109] = Sulfuras, Hand of Ragnaros, 5, 29
All Items[110] = Sulfuras, Hand of Ragnaros, 10, 0
All Items[111] = Sulfuras, Hand of Ragnaros, 10, 1
All Items[112] = Sulfuras, Hand of Ragnaros, 10, 2
All Items[113] = Sulfuras, Hand of Ragnaros, 10, 5
All Items[114] = Sulfuras, Hand of Ragnaros, 10, 29
All Items[115] = Sulfuras, Hand of Ragnaros, 2, 0
All Items[116] = Sulfuras, Hand of Ragnaros, 2, 1
All Items[117] = Sulfuras, Hand of Ragnaros, 2, 2
All Items[118] = Sulfuras, Hand of Ragnaros, 2, 5
All Items[119] = Sulfuras, Hand of Ragnaros, 2, 29
All Items[120] = Sulfuras, Hand of Ragnaros, 4, 0
All Items[121] = Sulfuras, Hand of Ragnaros, 4, 1
All Items[122] = Sulfuras, Hand of Ragnaros, 4, 2
All Items[123] = Sulfuras, Hand of Ragnaros, 4, 5
All Items[124] = Sulfuras, Hand of Ragnaros, 4, 29
